File: src/main/java/org/warcbase/pig/piggybank/ExtractBoilerpipeText.java
package org.warcbase.pig.piggybank;

import java.io.IOException;
import org.apache.commons.lang.StringUtils;

import org.apache.pig.EvalFunc;
import org.apache.pig.data.Tuple;
import de.l3s.boilerpipe.extractors.DefaultExtractor;
// Could also use tika.parser.html.BoilerpipeContentHandler, which uses older version of boilerpipe

/**
 * UDF for extracting raw text content from an HTML page, minus "boilerplate"
 * content (using boilerpipe).
 */
public class ExtractBoilerpipeText extends EvalFunc<String> {
  public String exec(Tuple input) throws IOException {
    if (input == null || input.size() == 0 || input.get(0) == null) {
      return null;
    }

    try {
      // Other available extractors: https://boilerpipe.googlecode.com/svn/trunk/boilerpipe-core/javadoc/1.0/de/l3s/boilerpipe/extractors/package-summary.html
      String text = DefaultExtractor.INSTANCE.getText((String) input.get(0)).replaceAll("[\\r\\n]+", " ").trim();
      if (text.isEmpty())
        return null;
      else
        return text;
    } catch (Exception e) {
      throw new IOException("Caught exception processing input row ", e);
    }
  }
}

File: src/main/scala/org/warcbase/spark/matchbox/ExtractBoilerpipeText.scala
package org.warcbase.spark.matchbox

import java.io.IOException
import de.l3s.boilerpipe.extractors.DefaultExtractor

/**
 * UDF for extracting raw text content from an HTML page, minus "boilerplate"
 * content (using boilerpipe).
 */
object ExtractBoilerpipeText {
  def apply(input: String) = {
    if (input.isEmpty) Nil
    else
      try {
        val text = DefaultExtractor.INSTANCE.getText(input).replaceAll("[\\r\\n]+", " ").trim()
        if (text.isEmpty) Nil
        else text
      } catch {
        case e: Exception =>
          throw new IOException("Caught exception processing input row " + e)
      }
  }
}


